v1/ast/oracle/matcher.go

```diff
@@ -0,0 +1,210 @@
+	return func(node ast.Node, compiler *ast.Compiler, parsed *ast.Module, result *matchResult) {
+		if t.isRef {
+			matchRef(t, node, compiler, parsed, result)
+		} else if t.isVar {
+			matchVar(t, node, compiler, parsed, result)
+		}
+	}
+}
+
+// matchRef searches for reference definitions in a node.
+// Ref usage sites are never definition sites (definitions come from rules/imports).
+func matchRef(t *target, node ast.Node, compiler *ast.Compiler, parsed *ast.Module, result *matchResult) {
+	term, ok := node.(*ast.Term)
+	if !ok {
+		return
+	}
+
+	nodeRef, ok := term.Value.(ast.Ref)
+	if !ok {
+		return
+	}
+
+	targetRef, ok := t.term.Value.(ast.Ref)
+	if !ok {
+		return
+	}
+
+	// no self defn
+	if !nodeRef.Equal(targetRef) {
+		return
+	}
+
+	// check compiler for rule definitions
+	if compiler != nil {
+		if loc := findRulesDefinition(compiler, nodeRef); loc != nil {
+			result.addRef(loc)
+		}
+	}
+
+	// check imports for prefix matches
+	if parsed != nil {
+		prefix := nodeRef.ConstantPrefix()
+		for _, imp := range parsed.Imports {
+			if path, ok := imp.Path.Value.(ast.Ref); ok {
+				if prefix.HasPrefix(path) {
+					result.addRef(imp.Path.Location)
+				}
+			}
+		}
+	}
+}
+
+// matchVar searches for variable definitions in a node.
+// Variables can be declared in-place (e.g., function args, iteration vars),
+// so targetLocation is used to skip self-definition at the declaration site.
+func matchVar(t *target, node ast.Node, compiler *ast.Compiler, parsed *ast.Module, result *matchResult) {
+	v, ok := t.term.Value.(ast.Var)
+	if !ok {
+		return
+	}
+
+	targetLocation := t.term.Location
+
+	switch n := node.(type) {
+	case *ast.Expr:
+		// Check for iteration variables in Every/SomeDecl
+		if every, ok := n.Terms.(*ast.Every); ok {
+			// For domains that are refs, look them up in the compiler
+			if domain, ok := every.Domain.Value.(ast.Ref); ok && compiler != nil {
+				if loc := findRulesDefinition(compiler, domain); loc != nil {
+					result.addRef(loc)
+				}
+			}
+			// Check iteration variables
+			result.addVarIfNotSelf(every.Key, v, targetLocation)
+			result.addVarIfNotSelf(every.Value, v, targetLocation)
+		} else if someDecl, ok := n.Terms.(*ast.SomeDecl); ok {
+			if len(someDecl.Symbols) > 0 {
+				if call, ok := someDecl.Symbols[0].Value.(ast.Call); ok && len(call) >= 2 {
+					// For domains that are refs, look them up in the compiler
+					domain := call[len(call)-1]
+					if domainRef, ok := domain.Value.(ast.Ref); ok && compiler != nil {
+						if loc := findRulesDefinition(compiler, domainRef); loc != nil {
+							result.addRef(loc)
+						}
+					}
+					// Check iteration variables
+					iterVars := call[1 : len(call)-1]
+					for _, iterVar := range iterVars {
+						result.addVarIfNotSelf(iterVar, v, targetLocation)
+					}
+				}
+			}
+		}
+	case *ast.Rule: // check function arguments
+		for _, arg := range n.Head.Args {
```

Comment:
Do we also need to check the rule's ref for var matches?
In most cases, these would however be considered as output, not input.

Reply:
Have opened https://github.com/open-policy-agent/opa/issues/8137 for future in case we want to do this